### DISCUSSION TITLE
Compile to WASM

### DISCUSSION THREAD

**Discussion by kitepad** [c-smile/sciter-js-sdk#150] [Ideas]
1. Sciter.js engine can compile to WASM.
2. Sciter.js app can compile to WASM. Or Quark can package app as WASM binary.

Benefit is :  Single source code  for multi-platform, developer don't need to maintain different html/css syntax for ONE ui.

**Comment by c-smile** [comment]
Why do you need WASM support built into the engine?

You can always embed existing WASM VM library into your application if you so need that. And there are [plenty of them](https://github.com/appcypher/awesome-wasm-runtimes).  You can link WASM VM statically into your app or put it as DLL - sciter's extension so it can be used in Quark / scapp context.

Normally Sciter use cases do not require such feature as applications can provide native functions for script easily.

**Reply by kitepad** [reply]
I’m not sure if the engine is needed, when my application is compiled to wasm and running in browser. And my original ideas is not embed WASM vm into my app, My expectations are:
1. single source (sciter.js syntax for html/css/js) for my app. I don't like to maintain different syntax for html tag, css selector and layout(flow/flex/grid).
2. my source can compile into wasm
3. my app's wasm can be loaded by browser 
4. can package my source into single executable using quark or wasm bin can be loaded by scapp.exe.

**Reply by c-smile** [reply]
> my source can compile into wasm

What is the "source" here? Some language, which one?

**Reply by kitepad** [reply]
My sciter.js app used html/css/js. For example, my demo app only have one file : demo.htm, but css style and js are embed in this file. So "source" mean demo.htm file.

**Reply by c-smile** [reply]
> My sciter.js app used html/css/js.

And where is WASM here? Or do you want sciter engine itself to be compiled as WASM ?

**Reply by kitepad** [reply]
I think sciter engine is app container like browser, my app is content view and developed by developer. Content view assets can be compiled as WASM binary (self contained). If this WASM binary can be run in browser, sciter engine don't need to be compiled. BUT if not, sciter engine has to be compiled as WASM.  
My idea is from egui (https://github.com/emilk/egui) , a UI toolkit for Rust.  I don't know if this is possible for sciter.js.
